### PR TITLE
Stage2 riscv64 support

### PR DESCRIPTION
**Pull request by pfgithub**
```zig
export fn _start() noreturn {
    print();
    exit();
}
fn print() void {
    asm volatile ("ecall"
        :
        : [sys] "{a7}" (64),
          [arg1] "{a0}" (1),
          [arg2] "{a1}" (@ptrToInt("Hello, World!\n")),
          [arg3] "{a2}" ("Hello, World!\n".len)
        : "memory"
    );
}
fn exit() noreturn {
    asm volatile ("ecall"
        :
        : [sys] "{a7}" (94),
          [arg1] "{a0}" (0)
        : "memory"
    );
    unreachable;
}
```
(`{x0}`…`{x31}` also work as register names)

Still need to do:
- [ ] save and reload `ra` in functions
- [x] align all instructions to 2 bytes
- [x] a test

### CODE REVIEW

**Comment by FireFox317** [comment]
Are you sure all these `packed` structs work as intented? We have bugs with `packed` structs in stage1 atm. related #3133 #2627

**Comment by pfgithub** [comment]
@FireFox317 I didn't expect them to be working properly, but they seem to be working and are the correct size. This outputs correctly:
```zig
test "" {
    const Stru = packed struct {
        opcode: u7,
        rd: u5,
        unused: u3 = 0,
        rsi1: u5,
        imm11: u11,
        signextend: u1 = 0,
    };
    @import("std").debug.warn("\n\n{b:0>32}\n", .{@intCast(u32, 0b01100000000100001000000010000001)});
    @import("std").debug.warn("{b:0>32}\n\n", .{@bitCast(u32, Stru{ .imm11 = 0b11000000001, .rsi1 = 1, .rd = 1, .opcode = 1 })});
}
```

File: src-self-hosted/codegen.zig [hunk 1]
```diff
@@ -1421,6 +1497,78 @@ fn Function(comptime arch: std.Target.Cpu.Arch) type {
...
+                    .ptr_embedded_in_code => unreachable,
+                    .unreach, .none => return, // Nothing to do.
+                    .undef => {
+                        if (!self.wantSafety())
+                            return; // The already existing value will do just fine.
+                        // Write the debug undefined value.
+                        switch (reg.size()) {
+                            64 => return self.genSetReg(src, reg, .{ .immediate = 0xaaaaaaaaaaaaaaaa }),
+                            else => unreachable,
+                        }
```
Comment (andrewrk) [inline src-self-hosted/codegen.zig:1512]: No need for this switch right? If all registers are always 64 bits.

**Comment by pixelherodev** [comment]
The size and to64 functions should probably be removed too. Those are
only needed for x64, they don't represent a concrete interface.

-- Email domain proudly hosted at https://migadu.com

**Comment by andrewrk** [comment]
I'll make a change in master branch regarding the to64 things that should clear up how to (not have to) do it for other architectures

**Comment by andrewrk** [comment]
Nice work! Merged in e4eb4396c27010d8d07b6fc73e7c7557170f06ad. In that branch before merging I also made `-Denable-qemu` work.

...actually I forgot to check that the test was working, looks like it's not failing when I mangle the "hello world" text. Will push a follow-up commit shortly. But anyway the code is merged into master.

**Comment by andrewrk** [comment]
Fixed in d61a9e37ae8f140407d1369500d21efbe2b198ab:

```
Test [1/1] test "self-hosted"...tests [1/19] call undefined local [1/1] update [2/3] parse/analysis/codtests [7/19] hello world [1/1] update [4/4] update index 0, mismatched stdout
====Expected (len=14):====
Hello, World!

====Actual (len=14):====
Hello2, World!
========

/home/andy/dev/zig/src-self-hosted/test.zig:650:40: 0x28042b in TestContext.runOneCase (test)
                        std.debug.panic(
                                       ^
/home/andy/dev/zig/src-self-hosted/test.zig:421:32: 0x27b1df in TestContext.run (test)
            try self.runOneCase(std.testing.allocator, &prg_node, case);
                               ^
/home/andy/dev/zig/src-self-hosted/test.zig:21:16: 0x27ad25 in test "self-hosted" (test)
    try ctx.run();
               ^
/home/andy/dev/zig/lib/std/special/test_runner.zig:48:28: 0x4a1f25 in std.special.main (test)
        } else test_fn.func();
                           ^
/home/andy/dev/zig/lib/std/start.zig:252:37: 0x27b8bd in std.start.posixCallMainAndExit (test)
            const result = root.main() catch |err| {
                                    ^
/home/andy/dev/zig/lib/std/start.zig:123:5: 0x27b5ff in std.start._start (test)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^

Tests failed. Use the following command to reproduce the failure:
/home/andy/dev/zig/zig-cache/o/72BKgrzz0IQ1aLeRaOFoFJ2h6DfavwqGEtek3FNxRhMTbh4kyBjTG4V7SQ2nFC0Y/test
The following command exited with error code 255:
/home/andy/dev/zig/build-release/zig test /home/andy/dev/zig/src-self-hosted/test.zig --pkg-begin build_options /home/andy/dev/zig/zig-cache/test_build_options.zig --pkg-end --cache-dir /home/andy/dev/zig/zig-cache --name test --pkg-begin stage2_tests /home/andy/dev/zig/test/stage2/test.zig --pkg-end 

Build failed. The following command failed:
/home/andy/dev/zig/zig-cache/o/Vir_I3dkRZtyxYXk0CWGKRTdft2RMrYfZnRBu3welLglJILnaN2Sv46SMzG_Sd9M/build /home/andy/dev/zig/build-release/zig /home/andy/dev/zig /home/andy/dev/zig/zig-cache test-stage2 -Denable-qemu

```

**Pull request closed without merging**